src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -1120,6 +1131,144 @@ protected static ImmutableArray<string> CheckTupleElementNames(int cardinality,
+        /// reasons for accessibility failure.</para>
+        /// </remarks>
+        public bool IsSymbolAccessibleWithin(
+            ISymbol symbol,
+            ISymbol within,
+            ITypeSymbol throughType = null)
+        {
+            if (symbol is null)
+            {
+                throw new ArgumentNullException(nameof(symbol));
+            }
+
+            if (within is null)
+            {
+                throw new ArgumentNullException(nameof(within));
+            }
+
+            if (!(within is INamedTypeSymbol || within is IAssemblySymbol))
+            {
+                throw new ArgumentException(CodeAnalysisResources.IsSymbolAccessibleBadWithin, nameof(within));
+            }
+
+            checkInCompilationReferences(symbol, nameof(symbol));
+            checkInCompilationReferences(within, nameof(within));
+            if (!(throughType is null))
+            {
+                checkInCompilationReferences(throughType, nameof(throughType));
+            }
+
+            return IsSymbolAccessibleWithinCore(symbol, within, throughType);
+
+            void checkInCompilationReferences(ISymbol s, string parameterName)
+            {
+                var containingAssembly = computeContainingAssembly(s);
+                if (!assemblyIsInReferences(containingAssembly))
+                {
+                    throw new ArgumentException(string.Format(CodeAnalysisResources.IsSymbolAccessibleWrongAssembly, parameterName), parameterName);
+                }
+            }
+
+            bool assemblyIsInReferences(IAssemblySymbol a)
+            {
+                if (assemblyIsInCompilationReferences(a, this))
+                {
+                    return true;
+                }
+
+                if (this.IsSubmission)
+                {
+                    // Submissions can reference symbols from previous submissions and their referenced assemblies, even
+                    // though those references are missing from this.References. We work around that by digging in
+                    // to find references of previous submissions. See https://github.com/dotnet/roslyn/issues/27356
+                    for (Compilation c = this.PreviousSubmission; c != null; c = c.PreviousSubmission)
+                    {
+                        if (assemblyIsInCompilationReferences(a, c))
+                        {
+                            return true;
+                        }
+                    }
+                }
+
+                return false;
+            }
+
+            bool assemblyIsInCompilationReferences(IAssemblySymbol a, Compilation compilation)
+            {
+                if (a == compilation.Assembly)
+                {
+                    return true;
+                }
+
+                foreach (var reference in compilation.References)
+                {
+                    var assembly = compilation.GetAssemblyOrModuleSymbol(reference) as IAssemblySymbol;
```

Comment:
>compilation.GetAssemblyOrModuleSymbol(reference) [](start = 35, length = 48)

Consider just directly comparing the output to `a` and avoiding the trycast altogether.